text/0000-raw-reference-operator.md

```diff
@@ -0,0 +1,189 @@
+to references, that have to *always hold*.  (This is not finally decided yet,
+but true in practice because of annotations we emit to LLVM.  It is also the
+next topic of discussion in the
+[Unsafe Code Guidelines](https://github.com/rust-rfcs/unsafe-code-guidelines/).)
+In particular, references must be aligned and dereferencable, even when they are
+created and never used.
+
+One consequence of these rules is that it becomes essentially impossible to
+create a raw pointer pointing to an unaligned struct field: `&packed.field as
+*const _` creates an immediate unaligned reference, triggering undefined
+behavior because it is not aligned.  Similarly, `&(*raw).field as *const _` is
+not just computing an offset of the raw pointer `raw`, it also asserts that the
+intermediate shared reference is aligned and dereferencable.  In both cases,
+that is likely not what the author of the code intended.
+
+To fix this, we propose to introduce a new primitive operation on the MIR level
+that, in a single MIR statement, creates a raw pointer to a given place.  No
+intermediate reference exists, so no invariants have to be adhered to.  We also
+add a lint for cases that seem like the programmer wanted a raw reference, not a
+safe one, but did not use the right syntax.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When working with unaligned or potentially dangling pointers, it is crucial that
+you always use raw pointers and not references: References come with guarantees
+that the compiler assumes are always upheld, and these guarantees include proper
+alignment and not being dangling.  Importantly, these guarantees must be
+maintained even when the reference is created and never used!  The following is
+UB (assuming `packed` is a variable of packed type):
+
+```rust
+let x = unsafe { &packed.field }; // `x` is not aligned -> undefined behavior
+```
+
+There is no situation in which the above code is correct, and hence it is a hard
+error to write this.  This applies to most ways of creating a reference, i.e.,
+all of the following are UB if `X` is not properly aligned and dereferencable:
+
+```rust
+fn foo() -> &T {
+  &X
+}
+
+fn bar(x: &T) {}
+bar(&X); // this is UB at the call site, not in `bar`
+
+let &x = &X; // this is actually dereferencing the pointer, certainly UB
+let _ = &X; // throwing away the value immediately changes nothing
+&X; // different syntax for the same thing
+
+let x = &X as &T as *const T; // this is casting to raw "too late"
+```
+
+The only way to create a pointer to an unaligned or dangling location without
+triggering undefined behavior is to *immediately* turn it into a raw pointer.
+All of the following are valid:
+
+```rust
+let packed_cast = unsafe { &packed.field as *const _ };
+let packed_coercion: *const _ = unsafe { &packed.field };
+let null_cast: *const _ = unsafe { &*ptr::null() } as *const _;
+let null_coercion: *const _ = unsafe { &*ptr::null() };
+```
```

Comment:
IIUC:

```rust
let packed_cast = &packed.field as *const _;
```

should also work. Might worth adding examples that do not use `unsafe` to take a reference to a packed struct field.

Reply:
The RFC mentions this (and a hard error even in unsafe blocks when not using `as`) as future possibilities building on top of this RFC. Are you suggesting they should be explicitly part of this RFC? I was trying to take baby steps here, not sure what is preferred.

Reply:
@RalfJung When the warning was added (Rust 1.24), the warning recommended that the correct way to fix this safe Rust code that compiled, run, and often worked correctly, was to just "use an `unsafe { ... }` block here instead". 

In Rust 1.29 the warning was changed to "This is always UB" (independently of whether this happens in an `unsafe` block or not). 

In fact, the RFC correctly argues below that if `&packed.field as *const _` constructs a raw pointer without creating a reference, then the code is _always_ safe. What value does requiring `unsafe` adds here? 

To me it seems that requiring `unsafe` for this is just a historical artifact of a poor decision.

Reply:
I agree with @gnzlbg that it would make sense for this to be safe. This would basically "fall out", I suppose, if we have a distinct MIR operation for it, since the unsafety checker is presently running after MIR construction.